tests/lax_scipy_sparse_test.py

```diff
@@ -193,6 +196,110 @@ def tree_unflatten(cls, aux_data, children):
+    A = rng(shape, dtype)
+    b = rng(shape[:1], dtype)
+    M = self._fetch_preconditioner(preconditioner, A, rng=rng)
+
+    def args_maker():
+      return A, b
+
+    self._CheckAgainstNumpy(
+        partial(scipy_bicgstab, M=M, maxiter=1),
+        partial(lax_bicgstab, M=M, maxiter=1),
+        args_maker,
+        tol=1e-5)
+
+    self._CheckAgainstNumpy(
+        partial(scipy_bicgstab, M=M, maxiter=2),
+        partial(lax_bicgstab, M=M, maxiter=2),
+        args_maker,
+        tol=1e-5)
+
+    self._CheckAgainstNumpy(
+        partial(scipy_bicgstab, M=M, maxiter=1),
+        partial(lax_bicgstab, M=M, maxiter=1),
+        args_maker,
+        tol=1e-5)
+
+    self._CheckAgainstNumpy(
+        np.linalg.solve,
+        partial(lax_bicgstab, M=M, atol=1e-6),
+        args_maker,
+        tol=1e-5)
+
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}_preconditioner={}".format(
+         jtu.format_shape_dtype_string(shape, dtype),
+         preconditioner),
+      "shape": shape, "dtype": dtype, "preconditioner": preconditioner}
+      for shape in [(2, 2), (7, 7)]
+      for dtype in float_types + complex_types
+      for preconditioner in [None, 'identity', 'exact']
+      ))
+  def test_bicgstab_on_identity_system(self, shape, dtype, preconditioner):
+    A = jnp.eye(shape[1], dtype=dtype)
+    solution = jnp.ones(shape[1], dtype=dtype)
+    rng = jtu.rand_default(self.rng())
+    M = self._fetch_preconditioner(preconditioner, A, rng=rng)
+    b = matmul_high_precision(A, solution)
+    tol = shape[0] * jnp.finfo(dtype).eps
+    x, info = jax.scipy.sparse.linalg.bicgstab(A, b, tol=tol, atol=tol,
+                                               M=M)
+    using_x64 = solution.dtype.kind in {np.float64, np.complex128}
+    solution_tol = 1e-8 if using_x64 else 1e-4
+    self.assertAllClose(x, solution, atol=solution_tol, rtol=solution_tol)
+
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name":
+       "_shape={}_preconditioner={}".format(
+         jtu.format_shape_dtype_string(shape, dtype),
+         preconditioner),
+      "shape": shape, "dtype": dtype, "preconditioner": preconditioner
+      }
+      for shape in [(2, 2), (4, 4)]
+      for dtype in float_types + complex_types
+      for preconditioner in [None, 'identity', 'exact']
+      ))
+  def test_bicgstab_on_random_system(self, shape, dtype, preconditioner):
+    rng = jtu.rand_default(self.rng())
+    A = rng(shape, dtype)
+    solution = rng(shape[1:], dtype)
+    M = self._fetch_preconditioner(preconditioner, A, rng=rng)
+    b = matmul_high_precision(A, solution)
+    tol = shape[0] * jnp.finfo(A.dtype).eps
+    x, info = jax.scipy.sparse.linalg.bicgstab(A, b, tol=tol, atol=tol, M=M)
+    using_x64 = solution.dtype.kind in {np.float64, np.complex128}
+    solution_tol = 1e-8 if using_x64 else 1e-4
+    self.assertAllClose(x, solution, atol=solution_tol, rtol=solution_tol)
+
```

Comment:
@shoyer By just adding this line, both GMRES and BICGSTAB seem to fail this test, moreso for the `complex64` and `float32` dtypes. But since `scipy` matches, I expect the same thing would happen if we had used `scipy` for the iterative solves.

Reply:
Try relaxing the error tolerance a little bit?

Reply:
Hmm, ok it's already 0.2 right now, and I saw this case for GMRES on my local tests:

```
Mismatch: 100%
Max absolute difference: 0.6610913
Max relative difference: 161.79356
 x: array(0.657005, dtype=float32)
 y: array(-0.004086, dtype=float32)
```

Reply:
Hmm! My guess is that there is a genuine bug in the derivative rule for GMRES. I would suggest commenting out this test for GMRES for now (opening an issue so we don't forget about it), and we can try to fix that in a follow-on PR.